py/server/deephaven/updateby.py

```diff
@@ -0,0 +1,276 @@
+#
+#     Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+#
+"""This module supports building various update-by clauses for use with the update-by operations on tables. """
+from enum import Enum
+from typing import Union, List
+
+import jpy
+
+from deephaven import DHError
+from deephaven._wrapper import JObjectWrapper
+from deephaven.jcompat import to_sequence
+
+_JUpdateByClause = jpy.get_type("io.deephaven.api.updateby.UpdateByClause")
+_JBadDataBehavior = jpy.get_type("io.deephaven.api.updateby.BadDataBehavior")
+_JEmaControl = jpy.get_type("io.deephaven.api.updateby.EmaControl")
+_JMathContext = jpy.get_type("java.math.MathContext")
+_JDateTimeUtils = jpy.get_type("io.deephaven.time.DateTimeUtils")
+
+
+class MathContext(Enum):
+    """An Enum for predefined precision and rounding settings in numeric calculation."""
+    UNLIMITED = _JMathContext.UNLIMITED
+    """unlimited precision arithmetic, rounding is half-up"""
+
+    DECIMAL32 = _JMathContext.DECIMAL32
+    """a precision setting matching the IEEE 754R Decimal32 format, 7 digits, rounding is half-even"""
+
+    DECIMAL64 = _JMathContext.DECIMAL64
+    """a precision setting matching the IEEE 754R Decimal64 format, 16 digits, rounding is half-even"""
+
+    DECIMAL128 = _JMathContext.DECIMAL128
+    """a precision setting matching the IEEE 754R Decimal128 format, 34 digits, rounding is half-evne"""
+
+
+class BadDataBehavior(Enum):
+    """An Enum defining ways to handle invalid data during EMA processing."""
+
+    RESET = _JBadDataBehavior.RESET
+    """Reset the state for the bucket to {@code null} when invalid data is encountered"""
+
+    SKIP = _JBadDataBehavior.SKIP
+    """Skip and do not process the invalid data without changing state"""
+
+    THROW = _JBadDataBehavior.THROW
+    """Throw an exception and abort processing when bad data is encountered"""
+
+    POISON = _JBadDataBehavior.POISON
+    """Allow the bad data to poison the result. This is only valid for use with NaN"""
+
+
+class EmaControl(JObjectWrapper):
+    """A EmaControl represents control parameters for performing EMAs (exponential moving average) with the table
+    update-by operation. """
+    j_object_type = _JEmaControl
```

Comment:
I'm trying to think a little down the line here.  Might `EmaControl` be too narrow of a name, especially if we are planning to add simple moving averages?  Do we expect SMA's to use the same control struct (which would need renaming to `MAControl` or such), or do we expect there to be a `SmaControl`?

You could imagine the current control structure being used for other kinds of averaging or aggregations.  Do we want to reuse it or have these other update-by-avgs implement their own control?

This may be a question for @lbooker42

Reply:
Created PR #2680 that renamed this object to `OperationControl`